Payment processor transaction parameter generation

rlucioni opened the pull request:
Includes a payment processor class for generating CyberSource transaction parameters. The order endpoint now returns CyberSource transaction parameters for newly created orders. Whether to use them is up to the LMS. This meets the requirements of [XCOM-114](https://openedx.atlassian.net/browse/XCOM-114).

@clintonb and @stephensanchez, could you please review this when you're able?

stephensanchez commented:
This looks really clean, Renzo. If we have 30 minutes today I'd love a walk through of this, though, because I'm not familiar with CyberSource. If able, maybe we can get @wedaly to review this quickly as well?

rlucioni commented:
@stephensanchez good idea, I should have thought to tag @wedaly. I'd be happy to walk you through it.

wedaly commented on extensions/payment/helpers.py:
```diff
@@ -0,0 +1,45 @@
+
+    Raises:
+        ImportError: If no module with the parsed module path exists.
+        AttributeError: If the module located at the parsed module path does not contain a
```
Nitpick: small typo here "a \n a"

rlucioni replied:
Fixed.

wedaly commented on extensions/settings/_oscar.py:
```diff
@@ -132,3 +132,12 @@
+
+
+########## PAYMENT PROCESSING
+PAYMENT_PROCESSOR_CLASS = 'extensions.payment.processors.Cybersource'
```
This isn't critical yet, but very soon the E-Commerce service is going to need to support multiple payment gateways (CyberSource and Paypal).  We might as well begin building the system with that assumption in mind.

Instead of defining a single payment processor class, I'd recommend generalizing this to a dictionary.  The keys of the dictionary can be processor names ("cybersource", "paypal", etc).  We'll need these names down the road when we begin defining receipt/cancellation/fulfillment end-points that understand the responses sent by each processor.

wedaly commented on extensions/api/views.py:
```diff
@@ -201,3 +204,13 @@ def _fulfill_order(self, order, shipping_event):
+        """Assemble a dictionary of metadata for the provided order."""
+        order_data = serializers.OrderSerializer(order).data
+
+        payment_processor = get_processor_class(getattr(settings, 'PAYMENT_PROCESSOR_CLASS'))
```
This is fine for now, but eventually we're going to need to extend the end-point so that it accepts an optional payment processor implementation (so we can have a button for PayPal in addition to the usual payment method).  Maybe have a setting to specify the default processor implementation?

wedaly commented on extensions/api/views.py:
```diff
@@ -127,9 +129,9 @@ def post(self, request):
             order = self._fulfill_order(order, shipping_event)
 
-        order_serializer = serializers.OrderSerializer(order)
+        order_data = self._assemble_order_data(order)
```
I think we're going to want to (a) save which payment processor we used for the order somewhere in the database, and (b) include info-level logging about which payment processor we used.  When we support multiple processors, we're going to need to know which order is paid for in which system.

clintonb replied:
I second this need. We should do this now to avoid data migrations in the future.

rlucioni replied:
All set here.

wedaly commented:
I'm realizing as I'm writing that supporting multiple payment processors is probably enough work to justify a separate story.  I'm okay with merging this PR now if you can create a JIRA ticket to track the work for extending this to support multiple gateways.

:+1:

clintonb commented on extensions/payment/errors.py:
```diff
@@ -0,0 +1,14 @@
+"""Exceptions and error messages used by payment processors."""
+class ProcessorError(Exception):
```
Not sure what other processors may come in the future; but, we can avoid future confusion by calling this `PaymentProcessorError`.

rlucioni replied:
Updated.

clintonb commented on extensions/payment/processors.py:
```diff
@@ -0,0 +1,181 @@
+class Cybersource(BasePaymentProcessor):
+    """CyberSource Secure Acceptance Web/Mobile (February 2015)
+
+    For reference, see http://goo.gl/KNXsK.
```
Link directly to http://apps.cybersource.com/library/documentation/dev_guides/Secure_Acceptance_WM/Secure_Acceptance_WM.pdf.

rlucioni replied:
Updated. What's wrong with the shortened URL?

clintonb replied:
There is nothing inherently wrong with it; but, unless we expect the URL to change frequently, there is no real need to have a proxy/shortened URL.

clintonb commented on extensions/payment/processors.py:
```diff
@@ -0,0 +1,181 @@
+    NAME = CS.NAME
+
+    def get_transaction_parameters(self, order, receipt_page_url=None, cancel_page_url=None, merchant_defined_data=None):
+        """Generate a dictionary of signed parameters CyberSource requires to complete a transaction.
```
Move this docstring to the base class? Willy pep8/pylint complain?

rlucioni replied:
This docstring includes details specific to CyberSource. I'll expand the base class' docstring to explain the purpose of the expected arguments. That should keep pep8/pylint happy, too. (Speaking of which, we need to tell Travis to check pep8/pylint.)

clintonb replied:
RE: Travis. We have a make target, validate, in Insights. It calls test_python and quality.

rlucioni replied:
Thanks for the tip.

clintonb commented on extensions/payment/processors.py:
```diff
@@ -0,0 +1,181 @@
+            ExcessiveMerchantDefinedData: If the provided merchant-defined data exceeds CyberSource's
+                optional field limit.
+        """
+        parameters = OrderedDict()
```
How does an `OrderDict` facilitate testing? I'm hesitant to change implementation to facilitate testing.

wedaly replied:
I believe the keys are joined into a comma separated list, which is used to generate the digital signature.  If the order of the keys is non-deterministic, then the signature might change, which could cause test failures.

I don't think this is only a testing concern though.  There's value in ensuring that the signatures we generate are 100% deterministic and reproducible :)

rlucioni replied:
@wedaly that's correct; @clintonb, I would have responded with the above.

clintonb commented on extensions/payment/processors.py:
```diff
@@ -0,0 +1,181 @@
+        in the same way they appear in 'signed_field_names'.
+
+        Arguments:
+            parameters (OrderedDict): A dictionary of transaction parameters.
```
This argument doesn't *need* to be an `OrderDict`.

rlucioni replied:
Good catch. Updated.

clintonb commented on extensions/payment/tests/test_processors.py:
```diff
@@ -0,0 +1,172 @@
+        logging.disable(logging.CRITICAL)
+
+        user = User.objects.create_user(
+            username='Gus', email='gustavo@lospolloshermanos.com', password='the-chicken-man'
```
Nice!

rlucioni replied:
:chicken:

wedaly replied:
pollos

clintonb commented on extensions/payment/tests/test_processors.py:
```diff
@@ -0,0 +1,172 @@
+        # Override all loggers, suppressing logging calls of severity CRITICAL and below
+        logging.disable(logging.CRITICAL)
+
+        user = User.objects.create_user(
```
Minor: Use django-dynamic-fixture (since we have it).

rlucioni replied:
But then I can't make users with clever names and emails, like Gus!

clintonb commented on extensions/payment/tests/test_processors.py:
```diff
@@ -0,0 +1,172 @@
+            parent=fried_chicken,
+            title=u'𝕃𝕠𝕤 ℙ𝕠𝕝𝕝𝕠𝕤 ℍ𝕖𝕣𝕞𝕒𝕟𝕠𝕤',
+            stockrecords__partner_sku=u'ṠÖṀЁṪḦЇṄĠ⸚ḊЁḶЇĊЇÖÜṠ',
+            stockrecords__price_excl_tax=D('9.99'),
```
Seems pretty expensive. Is this for a whole chicken?

rlucioni replied:
Based on the way I've configured the products, it looks like that's the price to buy the restaurant! Maybe that's how much you need to pay to open a franchise. Not such a bad deal, if you ask me.

clintonb commented on extensions/payment/tests/test_processors.py:
```diff
@@ -0,0 +1,172 @@
+
+    def setUp(self):
+        # Override all loggers, suppressing logging calls of severity CRITICAL and below
+        logging.disable(logging.CRITICAL)
```
Why no logging? Might be useful for debugging test failures.

rlucioni replied:
It's noisy. I figure breakpoints should be sufficient to debug test failures. If necessary, re-enabling logging while tests are running is easy.

clintonb commented on extensions/payment/tests/test_processors.py:
```diff
@@ -0,0 +1,172 @@
+class CybersourceTests(PaymentProcessorTests):
+    """Tests of the CyberSource processor class."""
+    UUID_HEX = 'madrigal'
+    PI_DAY = datetime.datetime(2015, 3, 14, 21, 26, 53)
```
`2015, 3, 14, 9, 26, 53`

rlucioni replied:
Technically, `2015, 3, 14, 21, 26, 53` is also correct: https://docs.python.org/2/library/datetime.html#datetime.datetime.hour.

I'll let you have this one, though.

clintonb replied:
Boo! 9AM!

clintonb commented on extensions/payment/tests/test_processors.py:
```diff
@@ -0,0 +1,172 @@
+
+        expected_parameters[CS.FIELD_NAMES.SIGNATURE] = sign(message, secret)
+
+        self.assertEqual(returned_parameters, expected_parameters)
```
I'm missing the need for an `OrderDict`.

rlucioni commented:
@wedaly I've created [XCOM-149](https://openedx.atlassian.net/browse/XCOM-149) to track the work for supporting multiple gateways.

clintonb commented:
My concerns regarding `OrderDict` have been addressed. Thanks. Overall, everything looks good; but, I would like to see the payment processor recorded in the database now, to avoid data migrations in the future. Data migrations aren't bad. I'm just hesitant to touch production data, especially commerce-related data.

rlucioni commented:
@wedaly and @clintonb, as suggested, I've added a `payment_processor` field to the order model, and extended the info-level log message emitted when orders are created to include information about the payment processor to be used.

I've also bumped the version of the `requests` module to 2.6.0, upgraded to fix a recently disclosed vulnerability.

rlucioni commented:
I've added a second commit to this PR which cleans up all pep8 and pylint violations in this project. I've also modified the Travis configuration to use a new `make` target which will run our unit tests in addition to running pep8 and pylint.

clintonb commented:
:+1:
